File: problems/Java/44. Wildcard Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        int slen = s.length(), plen = p.length();
        int sp = 0, pp = 0;
        int star = -1, mtch = 0;

        while (sp < slen) {
            if (pp < plen && (p.charAt(pp) == '?' || p.charAt(pp) == s.charAt(sp))) {
                sp++;
                pp++;
            } else if (pp < plen && p.charAt(pp) == '*') {
                star = pp;
                mtch = sp;
                pp++;
            } else if (star != -1) {
                pp = star + 1;
                mtch++;
                sp = mtch;
            } else {
                return false;
            }
        }

        while (pp < plen && p.charAt(pp) == '*') {
            pp++;
        }

        return pp == plen;
    }
}

File: problems/Python/44. Wildcard Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        slen, plen = len(s), len(p)
        sp, pp = 0, 0
        star = -1
        mtch = 0

        while sp < slen:
            if pp < plen and (p[pp] == '?' or p[pp] == s[sp]):
                sp += 1
                pp += 1
            elif pp < plen and p[pp] == '*':
                star = pp
                mtch = sp
                pp += 1
            elif star != -1:
                pp = star + 1
                mtch += 1
                sp = mtch
            else:
                return False

        while pp < plen and p[pp] == '*':
            pp += 1

        return pp == plen